stdlib/LibCURL.version

```diff
@@ -0,0 +1,2 @@
+LIBCURL_BRANCH = master
+LIBCURL_SHA1 = d018fe70928510b552600473e136de5d9bbe65b3
```

Comment:
Is this just the same as https://github.com/JuliaWeb/LibCURL.jl?

I think you need to at least change the UUID otherwise when someone uses an existing Manifest with that package they will immidiately get the one in the sysimage instead of the version they have in their manifest.

Reply:
That's what's inevitably going to happen with this. I've been saying that we need a way to have private snapshots of external packages for a while now. However, I'm first going to get this PR working and passing tests by any means necessary and then we can figure out what to do about all the f*¢&ing dependencies.

Reply:
> That's what's inevitably going to happen with this PR

Why inevitably? You could just change the UUID so it is not considered the same package. Manifests using the old package will use it, manifests using the new stdlib LibCURL will keep using that. Also, if we use the same UUID then we are effectively "hijacking" that package since it will no longer be updatable by its authors, post 1.6. The authors of LibCURL.jl might very well be fine with it but has there been any discussion about that?

> and others that we need to merge in 1.6

are there even more external packages that are going into stdlib for 1.6?

> we can figure out what to do about all the f*¢&ing dependencies.

Yeah, writing stdlibs sucks because you can't really use all the nice dependencies you want. But the strategy can't really be to work on a package externally, just like a normal package, decide that it should be an stdlib and then just "oh well, guess we have to put all its deps into stdlib as well"?

Reply:
What I mean by inevitably is that one of the following has to happen:

1. Rewrite `Downloader` to not use the full `LibCURL` wrapper, copy used methods
2. Add `LibCURL` as a stdlib, which prevents users from using it as a normal package
3. Add `LibCURL′` as a stdlib with a different UUID and a snapshot of the code (your suggestion)
4. Add a "private dependency" mechanism for doing 3 automatically and systematically

For now I'm just going with 2 but I agree that it's not a viable approach for a release, I just want to get this working with minimal fuss. Doing 3 is fine but it's pretty ugly and I think something like 4 would be better.

There's a general concept of a "private dependency" that can be extracted: if you want to use a package as a dependency but you promise that no objects of any types defined by that package will leak from your usage of it, then you can have a private copy of it. The general concept of public versus private dependencies is that there can only be a single public instance of a package but arbitrarily many private instances. All packages that have a public dependency on a package have to be able to agree on a version they can use, but if a package with a private dependency has a version conflict with another package with a public or private dependency on it, then they can just use a different copy with a different version.

This concept solves the issue of stdlibs wanting to use non-stdlib packages: as long as they use something as a private dependency, then it's fine. If something else has a public dependency on it that's compatible with the private copy that we've baked into the sysimg, then they can just use that one. If they need a different version, then we give them a different copy.

This also solves the very similar problem that VS Code, JuliaPro, etc. have: they need to have some packages (e.g. JSON) pre-loaded in the client process in order to function properly, but then they also want user code to be able to use whatever version of those packages they want. Currently this just doesn't work and if you're running code that wants to use certain versions of packages, they simply can't and their application might breaak. If they use these packages as private dependencies, then everything works fine: the IDE code uses the packages as private dependencies and the user code gets whatever version it wants.

I'm not sure we can get a feature like this into 1.6, but it seems like a really important concept that would solve many problems in the ecosystem in one fell swoop.